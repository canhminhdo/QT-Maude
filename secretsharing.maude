load quantum-circuits.maude

mod SECRETE-SHARING is
    pr CIRCUIT .

    vars S : Scalar .

    ops a b : -> Scalar .
    eq Pow(Abs(a), 2) .+ Pow(Abs(b), 2) = 1 .
    eq S .* Pow(Abs(a), 2) .+ S .* Pow(Abs(b), 2) = S .

    op initQState : -> Qubit .
    eq initQState = (q[0]: a . |0> + b . |1>) (q[1]: |0>) (q[2]: |0>) (q[3]: |0>) .

    op init : -> Config .
    eq init = {
        (qstate: initQState)
        (actions: (H(1) CX(1, 2) CX(1, 3) CX(0, 1) H(0) H(2) M(0) M(1) M(2) (c[1] == 1 ? X(3)) (c[0] == 1 ? Z(3)) (c[2] == 1 ? Z(3))))
        (prob: 1)
        (bits: empty)
        (isEnd: false)
    } .
endm

load model-checker .

mod SECRETE-SHARING-PREDS is
    pr SECRETE-SHARING .
    inc SATISFACTION .

    subsort Config < State .

    var OCs : Soup{OComp} .
    var PROP : Prop .
    var Q : Qubit .
    var Prob : Scalar .

    ***************************************************************************************
    *** Factor \gamma on a state for which |\gamma| = 1 is regarded as a 'global phase' ***
    *** States that differ only by a global phase are physically indistinguishable and  ***
    *** equivalent in all ways that are physically equivalent
    ***************************************************************************************
    op isSuccess : -> Prop .
    eq {(isEnd: true) (qstate: Q) (prob: Prob) OCs} |= isSuccess
    = Prob > 0 implies
    ((qubitAt(Q, 3) == qubitAt(qstate(init), 0) or (-1) . qubitAt(Q, 3) == qubitAt(qstate(init), 0))) .
    eq {OCs} |= PROP = false [owise] .
endm

mod SECRETE-SHARING-CHECK is
    inc SECRETE-SHARING-PREDS .
    inc MODEL-CHECKER .
    inc LTL-SIMPLIFIER .

    op ssProp : -> Formula .
    eq ssProp = <> isSuccess .
endm

in debug.maude

red modelCheck(init, ssProp) .

search in SECRETE-SHARING : init =>* {(qstate: Q:Qubit) (isEnd: true) (prob: P:Scalar) OCs:Soup{OComp}}
such that P:Scalar > 0 implies (qubitAt(Q:Qubit, 3) == qubitAt(qstate(init), 0) or (-1) . qubitAt(Q:Qubit, 3) == qubitAt(qstate(init), 0)) .

--- red (initQState).H(1) .
--- red (initQState).H(1).CX(1, 2) .
--- red (initQState).H(1).CX(1, 2).CX(1, 3) .
--- red (initQState).H(1).CX(1, 2).CX(1, 3).CX(0, 1) .
--- red (initQState).H(1).CX(1, 2).CX(1, 3).CX(0, 1).H(0) .