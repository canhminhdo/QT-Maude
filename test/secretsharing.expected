==========================================
reduce in SECRETE-SHARING-CHECK : modelCheck(init, ssProp) .
result Bool: true
==========================================
search in SECRETE-SHARING : init =>* {OCs:Soup{OComp} qstate: Q:Qubit prob:
    P:Scalar isEnd: true} such that (-1 . qubitAt(Q:Qubit, 3) == qubitAt(
    initState, 0) or qubitAt(Q:Qubit, 3) == qubitAt(initState, 0)) and P:Scalar
    > 0 = true .

OCs:Soup{OComp} --> bits: (0 |-> 0, 1 |-> 0, 2 |-> 0) actions: nil
Q:Qubit --> (q[0]: |0>) (q[1]: |0>) (q[2]: |0>) q[3]: a . |0> + b . |1>
P:Scalar --> 1/8

OCs:Soup{OComp} --> bits: (0 |-> 0, 1 |-> 0, 2 |-> 1) actions: nil
Q:Qubit --> (q[0]: |0>) (q[1]: |0>) (q[2]: |1>) q[3]: a . |0> + b . |1>
P:Scalar --> 1/8

OCs:Soup{OComp} --> bits: (0 |-> 0, 1 |-> 1, 2 |-> 0) actions: nil
Q:Qubit --> (q[0]: |0>) (q[1]: |1>) (q[2]: |0>) q[3]: a . |0> + b . |1>
P:Scalar --> 1/8

OCs:Soup{OComp} --> bits: (0 |-> 1, 1 |-> 0, 2 |-> 0) actions: nil
Q:Qubit --> (q[0]: |1>) (q[1]: |0>) (q[2]: |0>) q[3]: a . |0> + b . |1>
P:Scalar --> 1/8

OCs:Soup{OComp} --> bits: (0 |-> 0, 1 |-> 1, 2 |-> 1) actions: nil
Q:Qubit --> (q[0]: |0>) (q[1]: |1>) (q[2]: |1>) q[3]: (a .* -1) . |0> + (b .*
    -1) . |1>
P:Scalar --> 1/8

OCs:Soup{OComp} --> bits: (0 |-> 1, 1 |-> 0, 2 |-> 1) actions: nil
Q:Qubit --> (q[0]: |1>) (q[1]: |0>) (q[2]: |1>) q[3]: a . |0> + b . |1>
P:Scalar --> 1/8

OCs:Soup{OComp} --> bits: (0 |-> 1, 1 |-> 1, 2 |-> 0) actions: nil
Q:Qubit --> (q[0]: |1>) (q[1]: |1>) (q[2]: |0>) q[3]: a . |0> + b . |1>
P:Scalar --> 1/8

OCs:Soup{OComp} --> bits: (0 |-> 1, 1 |-> 1, 2 |-> 1) actions: nil
Q:Qubit --> (q[0]: |1>) (q[1]: |1>) (q[2]: |1>) q[3]: (a .* -1) . |0> + (b .*
    -1) . |1>
P:Scalar --> 1/8

No more solutions.
Bye.
